### ISSUE TITLE
Want to restore the latest version on iCloud

### ISSUE THREAD

**Issue by nakkisan** [vincode-io/Zavala#196]
I am having trouble with a document saved on iCloud.

I edited a document saved on iCloud on my iPhone for 10 minutes and then on my Mac for 2 hours.
When I then opened the document on my iPhone, the document reverted to the previous version that I edited on my iPhone, and **the edits I made on my Mac were deleted.**

Since I was creating a record of the meeting, it would be very difficult to restore the deleted portions.

So, is it possible to restore to the past version?

Translated with www.DeepL.com/Translator (free version)

**Comment by vincode-io** [comment]
I am so terribly sorry about this. This is a syncing collision bug that I just recently learned about. I am actually working on a fix for it right now.

Unfortunately, there is no back version. I might be able to do some data recovery for you as long as you can share your meeting information. The problem is the order of the top rows is lost, but if you have the raw rows, you can use your memory to put them back to together.

If you are interested in trying this, zip up the contents of this folder and send it to me.

~/Library/Group Containers/group.io.vincode.Zavala/Accounts/cloudKit

Also, let me know when you need the data. I have to write a custom program to attempt the recovery and it might take me several hours.

This is the worse thing that a programmer's mistake can do to a user. I’ll do anything to try to make it right by you.

-Maurice

**Comment by Wiiilmaa** [comment]
Is there any progress on this issue? It is killing my current se-cases and inhibits me from adopting zavala for other things.
Is there anything we could du to help fix the problem?
 Looking forward to having this issue solved,
  Jens Wilmer

**Comment by vincode-io** [comment]
Yes, I have been working on this. Unfortunately it is a very difficult problem to resolve because the iCloud API's that we use for syncing are worthless doing conflict resolution.

The problem comes down to doing a correct three way merge. You need the client, server, and ancestor data to determine which is the most correct. In the event of merge conflict, iCloud is supposed to provide all 3 records of data, at least according to the documentation. In fact, it will not provide the ancestor record. I had to open an incident ticket with Apple to find this out. An Apple engineer actually had to read the Apple's iCloud code to figure out that this was happening.

The bottom line is that I have to implement a way to recreate the ancestor record which is quite difficult to do efficiently. I know how to do it now, but it is a lot of work. That and personal issues have made getting this fixed take much, much longer than I or anyone would like.

Stay tuned. I will let everyone know when there is a test version of Zavala with the issue resolved so that it can be tested. In the mean time, I recommend always syncing Zavala on a device that you have been using it on before you switch to another one. Then sync on the next device before beginning editing. That avoids triggering the conflict resolution bug that causes data loss.

**Comment by vincode-io** [comment]
This issue has been fixed and will be available soon for testing via TestFlight as version 2.1.

I sincerely apologize for how long it has taken to resolve this bug. I'll eventually blog about it, but the bottom line it was a *lot* of work and I just wasn't able to devote that much time to Zavala during this last release cycle.
